packages/node-core/src/indexer/multiChainRewind.service.ts

```diff
@@ -0,0 +1,251 @@
+  Normal = 'normal',
+  /** The current chain is waiting for other chains to rewind. */
+  WaitOtherChain = 'waitOtherChain',
+  /** The current chain is executing rewind. */
+  Rewinding = 'rewinding',
+}
+export interface IMultiChainRewindService {
+  chainId: string;
+  status: RewindStatus;
+  waitRewindHeader?: Header;
+  setGlobalRewindLock(rewindTimestamp: number): Promise<void>;
+  /**
+   * Check if the height is consistent before unlocking.
+   * @param tx
+   * @param rewindTimestamp The timestamp to roll back to, in milliseconds.
+   * @returns the number of remaining rewind chains
+   */
+  releaseChainRewindLock(tx: Transaction, rewindTimestamp: number): Promise<number>;
+}
+
+export interface IMultiChainHandler {
+  processMultiChainRewind(rewindBlockPayload: MultiChainRewindPayload): void;
+}
+
+/**
+ * Working principle:
+ * multiChainRewindService is primarily responsible for coordinating multi-chain projects.
+ * When global.rewindLock changes, a PG trigger sends a notification, and all subscribed chain projects will receive the rollback notification.
+ * This triggers a rollback process, where the fetch service handles the message by clearing the queue.
+ * During the next fillNextBlockBuffer loop, if it detects the rewinding state, it will execute the rollback.
+ */
+@Injectable()
+export class MultiChainRewindService implements IMultiChainRewindService, OnApplicationShutdown {
+  private _status: RewindStatus = RewindStatus.Normal;
+  private _chainId?: string;
+  private dbSchema: string;
+  private rewindTriggerName: string;
+  private pgListener?: PoolClient;
+  private _globalModel?: PlainGlobalModel = undefined;
+  waitRewindHeader?: Header;
+  constructor(
+    private nodeConfig: NodeConfig,
+    private eventEmitter: EventEmitter2,
+    private sequelize: Sequelize,
+    private storeService: StoreService,
+    @Inject('IBlockchainService') private readonly blockchainService: IBlockchainService
+  ) {
+    this.dbSchema = this.nodeConfig.dbSchema;
+    this.rewindTriggerName = hashName(this.dbSchema, 'rewind_trigger', '_global');
+  }
+
+  private set chainId(chainId: string) {
+    this._chainId = chainId;
+  }
+
+  get chainId(): string {
+    assert(this._chainId, 'chainId is not set');
+    return this._chainId;
+  }
+
+  private set status(status: RewindStatus) {
+    this._status = status;
+  }
+
+  get status() {
+    assert(this._status, 'status is not set');
+    return this._status;
+  }
+
+  get globalModel() {
+    if (!this._globalModel) {
+      this._globalModel = new PlainGlobalModel(this.dbSchema, this.chainId, this.storeService.globalDataRepo);
+    }
+    return this._globalModel;
+  }
+
+  onApplicationShutdown() {
+    this.sequelize.connectionManager.releaseConnection(this.pgListener as Connection);
+  }
+
+  async init(chainId: string, reindex: (targetHeader: Header) => Promise<void>) {
+    this.chainId = chainId;
+
+    if (this.nodeConfig.multiChain) {
```

Comment:
For readability its a good idea to exit early and unnest where possible.

So here you can do
```ts
if (!this.nodeConfig.multiChain) {
    return;
}
```